lib/active_model_serializers/model.rb

```diff
@@ -3,42 +3,50 @@
 # serializable non-activerecord objects.
 module ActiveModelSerializers
   class Model
-    include ActiveModel::Model
     include ActiveModel::Serializers::JSON
+    include ActiveModel::Model
+
+    class_attribute :attribute_names
+    self.attribute_names = []
 
     def self.attributes(*names)
       attr_accessor(*names)
+      self.attribute_names = attribute_names | names.map(&:to_sym)
     end
 
-    attr_reader :attributes, :errors
+    attributes :id
+    attr_writer :updated_at
+
+    attr_reader :errors
 
     def initialize(attributes = {})
-      @attributes = attributes && attributes.symbolize_keys
       @errors = ActiveModel::Errors.new(self)
       super
     end
 
     # Defaults to the downcased model name.
     def id
-      attributes.fetch(:id) { self.class.name.downcase }
+      @id ||= self.class.name && self.class.name.downcase
     end
 
     # Defaults to the downcased model name and updated_at
     def cache_key
-      attributes.fetch(:cache_key) { "#{self.class.name.downcase}/#{id}-#{updated_at.strftime('%Y%m%d%H%M%S%9N')}" }
+      ActiveSupport::Cache.expand_cache_key([
+        self.class.name && self.class.name.downcase,
```

Comment:
`self.class.object_id` **is** a unique identifier for the class. For the the instances, I don't see any better solution than hashing the attributes, which would suck because it would either be complicated or fail for hashes. I'd vote for no caching for those models.